File: easy/206-Reverse-Linked-List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return None

        current = head # The current node
        prev = None # The previous nodes

        # While there is another node to go to
        while current.next: 
            nxt = current.next # Cache the next node
            current.next = prev # Set this node's next node to the previous node
            prev = current # Set the previous node to this node
            current = nxt # Set this node to the cached next node

        current.next = prev # Link last node to the rest of the nodes

        return current # Return the last node (stored in current)

File: easy/206-Reverse-Linked-List.c
#include <stdlib.h>

struct ListNode
{
    int val;
    struct ListNode *next;
};

// Reverses the linked list pointed to be head
struct ListNode *reverseList(struct ListNode* head)
{
    // Null list case
    if(head == NULL)    
    {
        return head;
    }

    struct ListNode *current = head;
    struct ListNode *next = current->next; 
    current->next = NULL;

    // While there are nodes left to be reversed
    while(next != NULL)
    {
        current = next; // Go the the next node
        next = current->next; // Store the node after the current node

        current->next = head; // Set the current node to point to head
        head = current; // Set the current node to the head 
    }

    // Return the head node 
    return head;
}